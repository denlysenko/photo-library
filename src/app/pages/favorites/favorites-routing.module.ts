import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';

import { FavoritesComponent } from './favorites.component';
import { FavoritesResolver } from './favorites.resolver';

const routes: Routes = [];

@NgModule({
  imports: [
    RouterModule.forChild([
      {
        path: '',
        component: FavoritesComponent,
        resolve: {
          photos: FavoritesResolver,
        },
      },
    ]),
  ],
  exports: [RouterModule],
})
export class FavoritesRoutingModule {}
