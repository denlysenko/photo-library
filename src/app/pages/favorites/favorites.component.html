<app-photo-grid [photos]="photos$ | async"
                (photoClicked)="onPhotoClick($event)"></app-photo-grid>
