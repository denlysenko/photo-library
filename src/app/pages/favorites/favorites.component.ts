import { ChangeDetectionStrategy, Component } from '@angular/core';
import { ActivatedRoute, Router } from '@angular/router';

import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';

import { Photo } from '../../interfaces';

@Component({
  selector: 'app-favorites',
  templateUrl: './favorites.component.html',
  styleUrls: ['./favorites.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class FavoritesComponent {
  readonly photos$: Observable<Photo[]> = this.route.data.pipe(
    map((data) => data.photos)
  );

  constructor(
    private readonly route: ActivatedRoute,
    private readonly router: Router
  ) {}

  onPhotoClick(photo: Photo): void {
    this.router.navigate(['/photos', photo.id]);
  }
}
