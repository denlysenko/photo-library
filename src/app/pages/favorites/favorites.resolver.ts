import { Injectable } from '@angular/core';
import { Resolve } from '@angular/router';

import { Observable } from 'rxjs';

import { PhotosService } from '../../data-access';
import { Photo } from '../../interfaces/photo.interface';

@Injectable()
export class FavoritesResolver implements Resolve<Photo[]> {
  constructor(private readonly photosService: PhotosService) {}

  resolve(): Observable<Photo[]> {
    return this.photosService.getAllFavorites();
  }
}
