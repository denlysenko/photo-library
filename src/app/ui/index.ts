export * from './header/header.module';
export * from './photo-tile/photo-tile.module';
export * from './photo-grid/photo-grid.module';
export * from './fullpage-spinner/fullpage-spinner.module';
export * from './infinite-scroll/infinite-scroll.module';
export * from './inline-spinner/inline-spinner.module';
