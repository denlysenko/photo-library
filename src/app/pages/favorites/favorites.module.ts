import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { PhotoGridModule } from '../../ui';
import { FavoritesRoutingModule } from './favorites-routing.module';
import { FavoritesComponent } from './favorites.component';
import { FavoritesResolver } from './favorites.resolver';

@NgModule({
  imports: [CommonModule, FavoritesRoutingModule, PhotoGridModule],
  declarations: [FavoritesComponent],
  providers: [FavoritesResolver],
})
export class FavoritesModule {}
