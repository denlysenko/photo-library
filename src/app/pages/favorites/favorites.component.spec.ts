import { CommonModule } from '@angular/common';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { ActivatedRoute, Router } from '@angular/router';

import { of } from 'rxjs';

import { PhotoGridModule } from '../../ui';
import { FavoritesComponent } from './favorites.component';

const photos = [
  {
    id: 'uuid',
    link: 'link',
  },
];

describe('FavoritesComponent', () => {
  let component: FavoritesComponent;
  let fixture: ComponentFixture<FavoritesComponent>;
  let router: Router;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [CommonModule, PhotoGridModule],
      declarations: [FavoritesComponent],
      providers: [
        {
          provide: ActivatedRoute,
          useValue: {
            data: of({
              photos,
            }),
          },
        },
        {
          provide: Router,
          useValue: {
            navigate: jest.fn(),
          },
        },
      ],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(FavoritesComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();

    router = TestBed.inject(Router);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should create', () => {
    expect(fixture).toMatchSnapshot();
  });

  describe('onPhotoClick', () => {
    it('should navigate to photo page', () => {
      fixture.debugElement.nativeElement
        .querySelector('[data-testid=photo-tile]')
        .click();

      expect(router.navigate).toHaveBeenLastCalledWith([
        '/photos',
        photos[0].id,
      ]);
    });
  });
});
