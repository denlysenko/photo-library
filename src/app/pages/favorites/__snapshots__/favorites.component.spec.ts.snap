// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoritesComponent should create 1`] = `
<app-favorites
  photos$={[Function Observable]}
  route={[Function Object]}
  router={[Function Object]}
>
  <app-photo-grid>
    <div
      class="photo-grid"
    >
      
      <app-photo-tile
        data-testid="photo-tile"
      >
        <div
          class="photo-tile card sm"
        >
          <img
            alt="photo"
            class="photo"
            src="link"
          />
        </div>
      </app-photo-tile>
    </div>
  </app-photo-grid>
</app-favorites>
`;
